### PR TITLE
Viewport Addon

### PR DESCRIPTION
**Pull request by shilman**
Issue: #1624 

## What @saponifi3d did 😉 
This is a fairly early PR of a viewport plugin which allows you to resize the iframe of storybook.

### Remaining Tasks
- [x] Fix lint issues
- [x] Add unit tests
- [x] Add documentation and screenshots to readme.

### Possible Features
- Responsive hooks, basically add the ability to drag the ends of the iframe to change the size
- Improve interaction with the `iframe` rather than just grabbing the iframe from the DOM add listeners to the iframe to allow us to interact with it

## Screen Cap
viewport

## How to test

**FIXME**

Is this testable with jest or storyshots?

Does this need a new example in the kitchen sink apps?

Does this need an update to the documentation?

If your answer is yes to any of these, please make sure to include it in your PR.

### CODE REVIEW

**Comment by saponifi3d** [comment]
i was able to fix-up the lint errors, add a readme, and wrote a bunch of unit tests (it's not 100% coverage - didn't think we needed tests for simple exports yet).

i also noticed the code coverage reports also include the `dist` directories from addons, want me to fix that in this PR, a separate one, or is someone already fixing that?

File: addons/viewport/.gitignore [hunk 1]
```diff
@@ -0,0 +1,5 @@
+*.sw*
```
Comment (ndelangen) [inline addons/viewport/.gitignore]: Please move any necessary gitignore to the root

File: addons/viewport/.storybook/addons.js [hunk 2]
```diff
@@ -0,0 +1 @@
+import '@storybook/addon-viewport/register';
```
Comment (ndelangen) [inline addons/viewport/.storybook/addons.js]: Please remove the dev setup for this addon, and use the cra-kitchen-sink to develop on.

File: addons/viewport/README.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
+# Storybook Viewport Addon
+
+Storybook Viewport Addon allows your stories to be displayed in different sizes and layouts in [Storybook](https://storybookjs.org).  This helps build responsive components inside of Storybook.
+
+This addon works with Storybook for: [React](https://github.com/storybooks/storybook/tree/master/app/react) and [Vue](https://github.com/storybooks/storybook/tree/master/app/vue).
```
Comment (ndelangen) [inline addons/viewport/README.md]: Awesome, great work!

File: addons/viewport/README.md [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+![Screenshot](docs/viewport.png)
```
Comment (ndelangen) [inline addons/viewport/README.md]: Can we make this is full URL? so it shows up in places like npm?

File: addons/viewport/README.md [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+## Installation
+
+Install the following npm module:
+
+    npm i -D @storybook/addon-viewport
```
Comment (ndelangen) [inline addons/viewport/README.md]: Use fenced code-blocks please like so:

```sh
npm i --save-dev @storybook/addon-viewport
```

File: addons/viewport/README.md [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+```js
+import '@storybook/addon-viewport/register'
+```
+
+This will register the Viewport Addon to Storybook and will show up in the action area.
+
+## FAQ
+
+#### Adding more devices
```
Comment (ndelangen) [inline addons/viewport/README.md]: minor detail, but I'd expect Frequently asked *Questions* to be actual questions 😃

File: addons/viewport/package.json [hunk 7]
```diff
@@ -0,0 +1,36 @@
+{
+  "name": "@storybook/addon-viewport",
+  "version": "3.2.0",
+  "description": "Storybook addon to change the viewport size to mobile",
+  "main": "dist/index.js",
+  "keywords": [
+    "storybook"
+  ],
+  "scripts": {
+    "build": "babel --source-maps --out-dir dist src",
```
Comment (ndelangen) [inline addons/viewport/package.json]: remove all but the `prepublish`

File: addons/viewport/package.json [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    "watch": "npm run build -- -w",
+    "prepublish": "npm run build && node ../../scripts/prepublish.js",
+    "storybook": "start-storybook -p 3000",
+    "deploy-storybook": "storybook-to-ghpages"
+  },
+  "license": "MIT",
+  "dependencies": {
+    "@storybook/react": "^3.2.5",
```
Comment (ndelangen) [inline addons/viewport/package.json]: Must be removed, this would install `@storybook/react` for vue users.

File: addons/viewport/package.json [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+    "prop-types": "^15.5.10"
+  },
+  "peerDependencies": {
+    "@storybook/addons": "^3.2.0",
+    "react": "*"
+  },
+  "devDependencies": {
+    "@storybook/addons": "^3.2.0",
+    "babel-cli": "^6.24.1",
```
Comment (ndelangen) [inline addons/viewport/package.json]: I think all the devDependencies can be removed.

should be using babel-preset-env if babel preset need to stay for some reason.

File: addons/viewport/src/components/Panel.jsx [hunk 10]
```diff
@@ -0,0 +1,120 @@
+/* global document */
```
Comment (ndelangen) [inline addons/viewport/src/components/Panel.jsx]: Please import any global from the `global` package.

This allows for easy mocking with jest and makes it very explicit.

File: addons/viewport/src/components/Panel.jsx [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+import { RotateViewport } from './RotateViewport';
+
+import * as styles from './styles';
+
+const storybookIframe = 'storybook-preview-iframe';
+const containerStyles = {
+  padding: 15,
+  width: '100%',
+  boxSizing: 'border-box',
+  fontFamily:
```
Comment (ndelangen) [inline addons/viewport/src/components/Panel.jsx]: Do not change the fonts please, or if you have to, you should get them from the `theme`-file from `lib/components`. That way we can ensure it's the same everywhere.

File: addons/viewport/src/components/Panel.jsx [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+
+  componentDidMount() {
+    this.iframe = document.getElementById(storybookIframe);
+  }
+
+  componentWillUnmount() {
+    this.props.channel.removeListener('addon:viewport:update', this.changeViewport);
+  }
+
+  iframe = undefined;
```
Comment (ndelangen) [inline addons/viewport/src/components/Panel.jsx]: I would prefer to set this in the constructor

File: addons/viewport/src/components/Panel.jsx [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+    }
+  };
+
+  toggleLandscape = () => {
+    const { isLandscape } = this.state;
+
+    this.setState({ isLandscape: !isLandscape }, this.updateIframe);
+  };
+
+  updateIframe = () => {
```
Comment (ndelangen) [inline addons/viewport/src/components/Panel.jsx]: arrow-functions cannot have a this binding, why are they object methods then?

If they are utility functions that are unrelated to the instance, we should hoist them to the top of the module. Or possibly even be moved into their own module.

File: addons/viewport/src/components/tests/Panel.spec.jsx [hunk 14]
```diff
@@ -0,0 +1,249 @@
+/* global document */
```
Comment (ndelangen) [inline addons/viewport/src/components/tests/Panel.spec.jsx]: Please import any global from the global package.

File: addons/viewport/src/components/tests/Panel.spec.jsx [hunk 15]
```diff
@@ -0,0 +1,249 @@
...
+import React from 'react';
+import { shallow } from 'enzyme'; // eslint-disable-line import/no-extraneous-dependencies
```
Comment (ndelangen) [inline addons/viewport/src/components/tests/Panel.spec.jsx]: We use a `<name>.test.js` naming convention for tests.

File: addons/viewport/src/components/tests/Panel.spec.jsx [hunk 16]
```diff
@@ -0,0 +1,249 @@
...
+        'addon:viewport:update',
+        subject.instance().changeViewport
+      );
+    });
+  });
+
+  describe('componentDidMount', () => {
+    let previousGet;
+
+    beforeEach(() => {
```
Comment (ndelangen) [inline addons/viewport/src/components/tests/Panel.spec.jsx]: low priority:

I prefer to use `beforeEach` and `afterEach` functions as last resort only.
The reason being it becomes harder and hard to know/understand all the steps executed to setup a test.

Not directly necessary to get this PR merged, but I'd like to see a `setup` function being used to get a shallow render in a specific state.

File: addons/viewport/src/components/tests/RotateViewport.spec.jsx [hunk 17]
```diff
@@ -0,0 +1,91 @@
+import React from 'react';
+import { shallow } from 'enzyme'; // eslint-disable-line import/no-extraneous-dependencies
```
Comment (ndelangen) [inline addons/viewport/src/components/tests/RotateViewport.spec.jsx]: Please use the .test.js naming convention here too

Comment (ndelangen) [reply addons/viewport/src/components/tests/RotateViewport.spec.jsx]: No use of .jsx file extention

File: addons/viewport/src/manager.jsx [hunk 18]
```diff
@@ -0,0 +1,24 @@
+import React from 'react';
+import addons from '@storybook/addons';
+
+import { Panel } from './components/Panel';
+
+const ADDON_ID = 'storybook-addon-viewport';
+const PANEL_ID = `${ADDON_ID}/addon-panel`;
+
+function addChannel(api) {
```
Comment (ndelangen) [inline addons/viewport/src/manager.jsx]: Looks like these could be arrow functions? any reason to make them function declarations?

File: examples/cra-kitchen-sink/.storybook/addons.js [hunk 19]
```diff
@@ -4,3 +4,4 @@ import '@storybook/addon-events/register';
 import '@storybook/addon-notes/register';
 import '@storybook/addon-options/register';
 import '@storybook/addon-knobs/register';
+import '@storybook/addon-viewport/register';
```
Comment (ndelangen) [inline examples/cra-kitchen-sink/.storybook/addons.js]: Great!

File: examples/vue-kitchen-sink/package.json [hunk 20]
```diff
@@ -10,6 +10,7 @@
     "@storybook/addon-centered": "^3.2.1",
     "@storybook/addon-notes": "^3.2.0",
     "@storybook/addon-knobs": "^3.2.0",
+    "@storybook/addon-viewport": "^3.2.0",
```
Comment (ndelangen) [inline examples/vue-kitchen-sink/package.json]: I think it needs to be registered in the addons.js too

File: examples/cra-kitchen-sink/.storybook/addons.js (hunk 19 above)
Comment (ndelangen) [reply examples/cra-kitchen-sink/.storybook/addons.js]: Also needs to be added as a dependency.

**Review by ndelangen — CHANGES_REQUESTED** [review]
If you need any help with making these changes, I'm very much willing to jump in and help! Just let me know. This feature is fantastic, a lot of people are going to be super happy with this!

File: addons/viewport/src/components/Panel.jsx (hunk 12 above)
Comment (Hypnosphi) [reply addons/viewport/src/components/Panel.jsx]: Is it a general suggestion not to use property initializers? Why? Maybe we should setup a eslint rule then?

File: addons/viewport/src/components/Panel.jsx (hunk 13 above)
Comment (Hypnosphi) [reply addons/viewport/src/components/Panel.jsx]: The function obviously uses `this`. Making it arrow function is just one more way to make it suitable as a callback: https://facebook.github.io/react/docs/handling-events.html ("If calling `bind` annoys you" section)

File: addons/viewport/README.md (hunk 6 above)
Comment (saponifi3d) [reply addons/viewport/README.md]: heh, fair. i'll update :)

File: addons/viewport/package.json (hunk 7 above)
Comment (saponifi3d) [reply addons/viewport/package.json]: 👍 i wasn't sure of the dev process prior. i'll cleanup so it only works in kitchen sink.

File: addons/viewport/src/components/tests/RotateViewport.spec.jsx (hunk 17 above)
Comment (saponifi3d) [reply addons/viewport/src/components/tests/RotateViewport.spec.jsx]: Should I change all `.jsx` files to be `.js`?

File: addons/viewport/src/manager.jsx (hunk 18 above)
Comment (saponifi3d) [reply addons/viewport/src/manager.jsx]: not a huge thing - but from what i understand chrome specifically will do a better job at optimizing named functions like this over arrow functions. happy to update to be an arrow function instead :D

File: addons/viewport/src/components/Panel.jsx (hunk 13 above)
Comment (saponifi3d) [reply addons/viewport/src/components/Panel.jsx]: @ndelangen are you saying to instead have this be a normal class function than in the constructor do `this.updateIframe = this.updateIframe.bind(this);`? Happy to move to that pattern instead if you want, just want to make sure i do it correctly :)

fwiw, this is a fairly common pattern as @Hypnosphi pointed out, but I think it's more important to be consistent through the codebase.

File: addons/viewport/src/components/tests/RotateViewport.spec.jsx (hunk 17 above)
Comment (ndelangen) [reply addons/viewport/src/components/tests/RotateViewport.spec.jsx]: Yes please!

**Comment by ndelangen** [comment]
Looking good @saponifi3d, looks like we could merge this soon.

What do you think @UsulPro @Hypnosphi ?

**Review by Hypnosphi — APPROVED** [review]
LGTM

File: addons/viewport/src/components/tests/RotateViewport.spec.jsx (hunk 17 above)
Comment (ndelangen) [reply addons/viewport/src/components/tests/RotateViewport.spec.jsx]: If you can rename all `.spec.` to `.test.` and all `.jsx` to `.js` I think this is good to go!

Let me know if you need any help! 👍

**Comment by saponifi3d** [comment]
@ndelangen just updated all the file names, let me know if there's anything else you wanted done (a lot of the comments just go smushed from the filename change).

thanks so much for the review / feedback!!!

**Review by ndelangen — APPROVED** [review]

**Review by danielduan — APPROVED** [review]

File: addons/viewport/register.js [hunk 21]
```diff
@@ -0,0 +1,3 @@
+// NOTE: loading addons using this file is deprecated!
+// please use manager.js and preview.js files instead
+require('./manager');
```
Comment (shilman) [inline addons/viewport/register.js]: @ndelangen @mnmtanish tangential to this PR but do you know when `register` was deprecated? This is the first time I've seen it. And isn't it strange that all of our existing addons use a deprecated format (AFAIK)? Seems like something we should figure out?

File: addons/viewport/src/components/Panel.js [hunk 22]
```diff
@@ -0,0 +1,119 @@
...
+
+  componentDidMount() {
+    this.iframe = document.getElementById(storybookIframe);
+  }
+
+  componentWillUnmount() {
+    this.props.channel.removeListener('addon:viewport:update', this.changeViewport);
+  }
+
+  iframe = undefined;
```
Comment (shilman) [inline addons/viewport/src/components/Panel.js]: @saponifi3d define in constructor?

**Comment by shilman** [comment]
@saponifi3d made a minor suggestion but I think it looks great to release for alpha. I'd like to generalize it in a way that also supports something like https://github.com/storybooks/storybook/pull/1394 but I think we can do that in a later revision and it's better for us to get this out as part of the alpha. Looking forward to merging and putting out our first 3.3 alpha this week!

**Comment by avaly** [comment]
What's the differences between this new addon and [`storybook-addon-scissors`](https://github.com/PeterPanen/storybook-addon-scissors)?

File: addons/viewport/src/index.js [hunk 23]
```diff
@@ -0,0 +1 @@
+export { register } from './manager';
```
Comment (Hypnosphi) [inline addons/viewport/src/index.js]: `manager.js` don't have this export

File: addons/viewport/register.js (hunk 21 above)
Comment (Hypnosphi) [reply addons/viewport/register.js]: It's also weird that we actually recommend importing this file in addon's documentation

File: addons/viewport/package.json [hunk 24]
```diff
@@ -0,0 +1,22 @@
+{
+  "name": "@storybook/addon-viewport",
+  "version": "3.2.0",
+  "description": "Storybook addon to change the viewport size to mobile",
+  "main": "dist/index.js",
```
Comment (Hypnosphi) [inline addons/viewport/package.json]: This file is broken

Comment (Hypnosphi) [reply addons/viewport/package.json]: #2589 fixes that by removing the `index.js` file and using `register.js` (or should it be `manager.js`?) as `main`
